pkg/ovs/configjob.go

```diff
@@ -0,0 +1,119 @@
+
+	ovnclient "github.com/openstack-k8s-operators/ovn-operator/api/v1beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConfigJob - prepare job to configure ovn-controller
+func ConfigJob(
+	ctx context.Context,
+	h *helper.Helper,
+	k8sClient client.Client,
+	instance *v1beta1.OVS,
+	configHash string,
+	labels map[string]string,
+) (*batchv1.Job, error) {
+
+	runAsUser := int64(0)
+	privileged := true
+	jobTTLAfterFinished := int32(10)
+
+	dbmap, err := ovnclient.GetDBEndpoints(ctx, h, instance.Namespace, map[string]string{})
+	if err != nil {
+		return nil, err
+	}
+
+	envVars := map[string]env.Setter{}
+	envVars["KOLLA_CONFIG_FILE"] = env.SetValue(KollaConfigAPI)
+	envVars["KOLLA_CONFIG_STRATEGY"] = env.SetValue("COPY_ALWAYS")
+	envVars["CONFIG_HASH"] = env.SetValue(configHash)
+	envVars["OvnBridge"] = env.SetValue(instance.Spec.ExternalIDS.OvnBridge)
+	envVars["OvnRemote"] = env.SetValue(dbmap["SB"])
+	envVars["OvnEncapType"] = env.SetValue(instance.Spec.ExternalIDS.OvnEncapType)
+	envVars["OvnEncapIP"] = EnvDownwardAPI("status.podIP")
+	envVars["EnableChassisAsGateway"] = env.SetValue(fmt.Sprintf("%t", instance.Spec.ExternalIDS.EnableChassisAsGateway))
+	envVars["PhysicalNetworks"] = env.SetValue(getPhysicalNetworks(instance))
+	envVars["OvnHostName"] = EnvDownwardAPI("spec.nodeName")
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name + "-configuration",
+			Namespace: instance.Namespace,
+			Labels:    labels,
+		},
+		Spec: batchv1.JobSpec{
+			TTLSecondsAfterFinished: &jobTTLAfterFinished,
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					RestartPolicy:      "OnFailure",
+					ServiceAccountName: ServiceAccountName,
+					Containers: []corev1.Container{
+						{
+							Name:  instance.Name + "-ovn-configuration-sync",
+							Image: instance.Spec.OvnContainerImage,
+							Command: []string{
+								"/bin/bash", "-c",
+							},
+							Args: []string{
+								"/usr/local/bin/container-scripts/init.sh",
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Capabilities: &corev1.Capabilities{
+									Add:  []corev1.Capability{"NET_ADMIN", "SYS_ADMIN"},
+									Drop: []corev1.Capability{},
+								},
+								RunAsUser:  &runAsUser,
+								Privileged: &privileged,
+							},
+							Env:          env.MergeEnvs([]corev1.EnvVar{}, envVars),
+							VolumeMounts: GetOvnVolumeMounts(),
+							Resources:    instance.Spec.Resources,
+						},
+					},
+					Volumes: GetVolumes(instance.Name),
+				},
+			},
+		},
+	}
+
+	numberOfJobsNeeded, err := getOvsPodsNumber(
+		ctx,
+		k8sClient,
+		instance,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	job.Spec.Completions = &numberOfJobsNeeded
+	job.Spec.Parallelism = &numberOfJobsNeeded
+
+	if instance.Spec.NodeSelector != nil && len(instance.Spec.NodeSelector) > 0 {
+		job.Spec.Template.Spec.NodeSelector = instance.Spec.NodeSelector
```

Comment:
does the job need to run on the worker node where the ovs pod is running? I mean do you need to set the NodeSelector based on  ovs pod worker info, or can it run on whatever worker node is chosen for the job pod?

Reply:
ack, where do you make sure that the job pod runs on the same worker node as the ovs pod? I don't we set this right now, but maybe I have missed something

Reply:
I didn't think that would be guaranteed by the parallelism and completions. My understanding was that the cube-scheduler would ultimately decide where to schedule the pods. Given a clean environment, this would probably end up doing as you describe. But, with active workloads, I think we could end up with two pods on a single node in some circumstances. My understanding about this feature could be wrong. 

It might be a case where we need a bit more complex nodeAffinity rules:
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/

We could probably add some type of label to the nodes where openstack-ovs is deployed, then use nodeAffinity rules to ensure one pod per node is run or something. 

This is just based on my general understanding of jobs and not any real world testing. But I can spend some time playing with it if no one else has a more definitive answer.

Reply:
right, thats not how it is going to work. the pods can be scheduled wherever the workload best fits onto if you don't specify where it has to run. If this should be a separate job so that it can be done without restart of ovn controller, I think you'd have to created individual jobs, one per worker node using the nodeSelector in the pod definition where the hostname matches the running ovs pods worker or another method as explained in https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ 

~~~
  nodeSelector:
    kubernetes.io/hostname: <pod worker>
~~~

Another thing I was thinking about is if the PostStart hook could be used for that [1], but then if things change also ovn-controller gets restarted, which you want to address with this PR.
An alternative might be to use a 2nd daemonset and use the script it runs to only run it once until it succeeded on a worker node. 

[1] https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks

Reply:
Yeah, that's where I was going with labels. If we thought it would be acceptable for us to tag nodes we're using with a label (similar to what kubevirt does). Then we can use that label to define a nodeAffinity. But if we already have the names of the nodes we're running the pods on, it's easy enough to just iterate through them for this job creation.

Maybe `getOvsPodsNumber` needs to be `getOvsPodsNode` and return a slice of node names? Like: `[]string{"node1","node2","node3"}` instead?